### DISCUSSION TITLE
AIR for MacOS Instructions

### DISCUSSION THREAD

**Discussion by Bilbosax** [airsdk/Adobe-Runtime-Support#716] [Q&A]
I am wanting to release a desktop version of my AIR app for macOS.  I do not want to release it from their App Store, I just want people to be able to download it from my website and install it without macOS making any fuss for them.  I know that I will probably need some sort of paid certificates, and will need to go through a review with Mac, and will likely have to package some information in my AIR app.  But I have no idea how to do any of these things.  Can someone point me to a useful reference for releasing on MacOS, or at the very least, give me an outline of what I need to do?

Thanks!

**Comment by ajwfrost** [comment]
If you don't want it to cause any problems for Catalina/later versions then yes you'd need to have the Apple developer account and create Developer ID certificates for development and maybe installation. If you then use AIR SDK / ADT to create a "bundle", you would end with a .app file that's got the application structure with all the files in the right place. You then need to sign all the binaries e.g. with the commands below, and then there are various ways to package it up for deployment: either into a .pkg file, or into a .dmg file. And then you notarize the final structure: there are fairly good instructions on Apple's website for this, and when you get a failure the logs generally tell you what you did wrong..

So for code signing:
```
1.    find "MyApp.app" -name "*.framework" -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
2.    find "MyApp.app" -name "*.plugin" -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
3.    find "MyApp.app" -name "*.dylib" -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
4.    find "MyApp.app" -name "A2712Enabler" -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;
5.    codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v MyApp.app
```
The entitlements plist file should be perhaps:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
        <key>com.apple.security.cs.allow-jit</key>
        <true/>
        <key>com.apple.security.cs.allow-unsigned-executable-memory</key>
        <true/>
        <key>com.apple.security.cs.disable-library-validation</key>
        <true/>
</dict>
</plist>
```
which allows the ActionScript JIT capability and the fact that we turn normal memory into executable memory; disabling library validation may be something you get away without but we add it to our entitlements so that people can load in other libraries whilst developing..

Once you have something that's signed, try notarizing it (you can notarize an .app if you zip it up first) to see whether there are any binaries that escaped all the above code-sign requests...

If you then want to create a .pkg file, you can use the "pkgbuild" command and this can take an argument for signing it (I'm going from memory now..). Or you can create a .dmg that allows your application to be installed... you only need to then notarize the outermost 'container' i.e. the pkg or dmg file, you don't actually need to notarize the .app bundle. if you need more details on these parts let me know and I'll check with the guy here who had done this, my recollection gets a little vague..

thanks

**Reply by Bilbosax** [reply]
Thanks for all the information Andrew.  I will get back to you on whether I can get all of this to work.  Do you need to get a verisign or some type of paid certificate for Apple so that your app doesn't say that it comes from an untrusted source and posts some kind of warning about using it to the user?

**Reply by ajwfrost** [reply]
I believe so, but it needs to come from Apple rather than any third party.
https://developer.apple.com/documentation/xcode/notarizing_macos_software_before_distribution?language=objc
https://developer.apple.com/support/certificates/
You basically need a "Developer ID" certificate that you generate once you have an Apple developer account..

**Comment by daghendrik** [comment]
I'm getting the error message:

`MyApp.app/Contents/Frameworks/Adobe AIR.framework: bundle format is ambiguous (could be app or framework)` 

when running the first code signing command ( find "MyApp.app" -name "*.framework" -exec codesign --entitlements ent.plist -s 'Developer ID Application' -f -o runtime --strict --timestamp --deep -v {} \;) On Big Sur 11.2.3 , MacBook Pro 2015 x86_64

Any trick to overcome this error?

**Reply by ajwfrost** [reply]
Can you do
```
ls -la MyApp.app/Contents/Frameworks/Adobe\ AIR.framework
```
and check that it's set up with symbolic links there?
I have seen that error before but am not getting it now (Big Sur 11.0, Mac Mini M1, AIR 33.1.1.444)

thanks

**Reply by SUSDEV** [reply]
Not to be too self-centered but I'm attaching my detailed instructions for compiling on Catalina, the initial part having to fix some symbolic links in what AIR produces. I haven't tried 33.1.1.444 yet but the last time I compiled for production (Sept 2020) I did have to do some fix-ups. I am going to push a new build in 1-2 months so I will update with what I find.
[Building and Notarizing - No Passwords.txt](https://github.com/Gamua/Adobe-Runtime-Support/files/6290850/Building.and.Notarizing.-.No.Passwords.txt)

**Reply by daghendrik** [reply]
Fixing the symbolic links, as described in the instructions by @SUSDEV, did the trick. Notarization failed on my first try, but succeeded after fixing some files I had overlooked (that had to be signed or removed). Having runned the unsigned app first might also have had an impact (tip from https://github.com/tuarua).

**Reply by ajwfrost** [reply]
@daghendrik that's great - but could you please confirm what extra AIR SDK release version you were using, and what changes you needed to make? then we can ensure that it's built properly into the ADT tool, I thought we'd already got it sorted out properly following some discussions with @SUSDEV a while back...

**Reply by daghendrik** [reply]
Absolutely. For packaging and notarization of my app I did the following:

1. Cloned https://github.com/tuarua/WebViewANE/blob/master/mac_installer/sign_developerId.sh (Script by TuaRua)
2. Created a captive runtime bundle (.app) in IntelliJ for **Adobe AIR 32.0.0.116** (Haven't tested the latest Harman SDK yet)
3. Copied that .app into the "bin-release" folder, referenced in the sign_developerId.sh script
4. Modified the script with my credentials and information, as described [here](https://github.com/tuarua/WebViewANE/blob/master/mac_installer/README.md#packaging-a-notarized-app-for-self-distribution)
5. Modified the script (sign_developerId.sh) with the lines below, taken from @SUSDEV 's instructions, adding them just below the last "rm -r" line of the "Cleaning up unneeded AIR files"-section (in sign_developerId.sh) 
6. Executed  the script. 

```
echo ${cyn} "Fixing Symlinks" ${white}
cd "$APP.app/Contents/Frameworks/Adobe AIR.framework"
rm "Adobe AIR"
ln -s "Versions/Current/Adobe AIR" "Adobe AIR"
ln -s "Versions/Current/Adobe AIR_64" "Adobe AIR_64"
rm -fr "Resources/"
ln -s "Versions/Current/Resources" "Resources"
cd "Versions/"
rm -fr "Current"
ln -s "1.0" "Current"
cd "$pathtome"
xattr -cr $APP.app
```

**Reply by ajwfrost** [reply]
Okay perfect, thanks. Yes, those are the ones that we knew about and the latest SDK that we've got should set up those links properly..

**Comment by yishayw** [comment]
Does the adt 'native' target not work on mac? I am getting a dmg file but when trying to run it tells me 'this application requires an update to adobe air that is not available for your system'

**Reply by ajwfrost** [reply]
What version of Adobe AIR do you have on the system? If it's an Adobe one then it's likely to give the above error I guess, since Adobe's versions won't know about our updates.

**Reply by yishayw** [reply]
So, you're saying I need to uninstall the shared runtime if it's pre-33? I'm compiling using AIR 33.

**Reply by SUSDEV** [reply]
Until @ajwfrost answers officially I don't know if you saw the release notes for 33.1.1.533:

For developers who are packaging applications for desktop AIR, there is now a shared AIR runtime that is available for end users to download at https://airsdk.harman.com/runtime. However, we continue to recommend that applications are packaged up with the captive bundle mechanism to include the runtime and remove the dependency upon this shared package.

**Reply by ajwfrost** [reply]
If you're compiling using AIR 33 and creating an application that has a namespace of 33.1, then when you try to run this on a computer with only AIR 32 on it, AIR will see that it can't run the 33.1-based app. So then you'll get that message.

You may be able to create your app with a 32.0 namespace and build/package it with the AIR 33 SDK, and it should then still run on AIR 32.

Or, yes if you uninstall the AIR 32 shared runtime and then install the AIR 33 shared runtime, it should work. I'm a little hesitant in asking people to uninstall Adobe's versions as it's often not possible to get those back again in case you needed them.. so perhaps worth checking the other option first... although on a Mac, the Adobe version will be increasingly restricted as the Apple rules/updates continue to come out..

**Reply by yishayw** [reply]
>creating an application that has a namespace of 33.1

Where is that specified?

I must be missing something, but I would have thought using the 'native' option means the runtime is bundled in the application and therefore it doesn't matter what shared runtimes are installed on the computer.

Could it be that I first need to choose the 'bundled' option and the use that as the source for the 'native' option?

**Reply by ajwfrost** [reply]
Namespace is at the top of the AIR application descriptor file, although if you're using an IDE like Animate or similar, it's probably being set for you - look for a file called "something-app.xml" or "application.xml" and see if this has e.g.
```<application xmlns="http://ns.adobe.com/air/application/33.1">```
at the top. It's a little convoluted though as you'd also need to build the SWF with a version code of 43 or less. It's all to try to ensure backwards compatibility works properly....

But in terms of the 'native' option: for both Windows and Mac, this just creates a 'native' installer (.dmg or .msi) that will install an AIR application using an existing shared runtime. I think you're right though, and it would make sense to also include the runtime so that it's then a self-contained package, and this would also get rid of some of the other issues that people see with code-signing for the MacOS apps on recent OS versions..

For now: you can could use the 'bundle' target instead of the 'native' one: this would leave you with an .app bundle which you can then turn into a pkg file perhaps, or create a dmg file with a "copy to applications" sort of view.

**Reply by yishayw** [reply]
>this just creates a 'native' installer (.dmg or .msi) that will install an AIR application using an existing shared runtime

Ok, got it. That was where I was confused. I was hoping to use 'native' to work around the pgkbuild issues I'm having, but I guess I'll have to figure those issues separately.

Thanks to both of you for the quick responses.
